### ISSUE TITLE
Can't use `_typeshed` in TYPE_CHECKING block

### ISSUE THREAD

**Issue by nineteendo** [tox-dev/sphinx-autodoc-typehints#479]
Example:

```python
if TYPE_CHECKING:
    from _typeshed import StrPath

def foo(path: StrPath) -> StrPath:
    return path
```

Output (fails a build on read the docs):

```none
WARNING: Failed guarded type import with ModuleNotFoundError("No module named '_typeshed'")
```

Stubs for `_typeshed`: https://github.com/python/typeshed/tree/main/stdlib/_typeshed

**Comment by gaborbernat** [comment]
I do not see how we can solve this, but if someone has ideas PR welcome. I would say for now do not use `_typeshed` imports.

**Comment by nineteendo** [comment]
Yeah, unless this can be implemented without executing the code block?

**Comment by gaborbernat** [comment]
The code block needs to be executed, otherwise we wouldn't be able to resolve some of the objects.

**Comment by gaborbernat** [comment]
Validated that this no longer reproduces - using `_typeshed.StrPath` inside a `TYPE_CHECKING` block no longer produces "Failed guarded type import" or "No module named '_typeshed'" warnings. This was fixed through improvements to guarded import handling in #225 (73aa9b6) and #393 (3eeb664).

If you still encounter this, please open a new issue with a reproducer.
